### ISSUE TITLE
C# Roslyn "process terminated" NRE critical failure, with repro

### ISSUE THREAD

**Issue by mgravell** [dotnet/roslyn#72862]
It works in devenv (including the test runner), so I don't know at what point I introduced the problem, and it is a wide branch in a big codebase; sorry.

Tested with 8.0.200,  and 8.0.203 - repros on both; tested on 9.0.100-preview.1.24101.2 - no repro, so *this might already be fixed*; wanted to report it because it is so severe, and it might not have been *intentionally* fixed (and thus backported for next service release).

Short version:

```
git clone https://github.com/StackExchange/StackExchange.Redis.git
cd StackExchange.Redis
git checkout cs-bug
cd .\src\RESPite
dotnet build
```

After the restore, it will vomit on its own shoes:

``` txt
➜ dotnet build
MSBuild version 17.9.4+90725d08d for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\8.0.200\Roslyn\Microsoft.CSharp.Core.targets(84,5): error : Process terminated. System.NullReferenceException: Object reference not set to an instance of an object. [C:\Code\StackExchange.Redis\src\RESPite\RESPite.csproj::TargetFramework=net472]
C:\Program Files\dotnet\sdk\8.0.200\Roslyn\Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol.get_ParameterCount() [C:\Code\StackExchange.Redis\src\RESPite\RESPite.csproj::TargetFramework=net472]
C:\Program Files\dotnet\sdk\8.0.200\Roslyn\Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CSharp.InMethodBinder.LookupSymbolsInSingleBinder(LookupResult result, String name, Int32 arity, ConsList`1 basesBeingResolved, LookupOptions options, Binder originalBinder, Boolean diagnose, CompoundUseSiteInfo`1& useSiteInfo) [C:\Co
de\StackExchange.Redis\src\RESPite\RESPite.csproj::TargetFramework=net472]
C:\Program Files\dotnet\sdk\8.0.200\Roslyn\Microsoft.CSharp.Core.targets(84,5): error :    at Microsoft.CodeAnalysis.CSharp.Binder.LookupSymbolsInternal(LookupResult result, String name, Int32 arity, ConsList`1 basesBeingResolved, LookupOptions options, Boolean diagnose, CompoundUseSiteInfo`1& useSiteInfo) [C:\Code\StackExchange.Redis\src\RESPite\RE
SPite.csproj::TargetFramework=net472]

.... lots and lots and lots

[C:\Code\StackExchange.Redis\src\RESPite\RESPite.csproj::TargetFramework=net461]
    0 Warning(s)
    1008 Error(s)
```

**Comment by jaredpar** [comment]
@jcouv PTAL. The stack doesn't immediately look like any of the regressions we recently addressed.

**Comment by jcouv** [comment]
TL;DR: this was fixed by a refactoring PR that reduce reentrancy when initializing event accessor symbols.

In `Debug` mode, this repro'es as an assertion in `SourceEventAccessor.Parameter`. Looking at the history for this file, I found a related change: https://github.com/dotnet/roslyn/pull/71497
I verified that this PR fixed the issues (crash in `Release` mode, assertion in `Debug` mode) by running the repro with the compiler at that commit vs. the compiler at the preceding commit.
